### PR TITLE
Add metronome destination

### PR DESCRIPTION
**Pull request by cozmo**
Initial WIP of the metronome destination. Once this PR is approved it'll be made against the upstream segment repo, and then deployed as a destination our users can add.

### CODE REVIEW

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Send Event',
+  description: 'Send an event to Metronome',
+  fields: {
+    transaction_id: {
+      type: 'string',
+      label: 'transaction_id',
+      // TODO: copy
+      description: 'The Metronome transaction id uniquely indentifies an event to ensure Metronome only processes each event once.',
```
Comment (smarx) [inline packages/destination-actions/src/destinations/metronome/sendEvent/index.ts]: Capitalizing ID and fixing a typo:

```suggestion
      description: 'The Metronome transaction ID uniquely identifies an event to ensure Metronome only processes each event once.',
```

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            }
+          }
+        }
+      }
+    },
+    timestamp: {
+      type: 'datetime',
+      label: 'timestamp',
+      // TODO: copy
+      description: 'The timestamp Metronome should use for this event.',
```
Comment (smarx) [inline packages/destination-actions/src/destinations/metronome/sendEvent/index.ts]:
```suggestion
      description: 'The timestamp at which this event occurred.',
```

**Comment by cozmo** [comment]
@smarx requested your review on copy in `packages/destination-actions/src/destinations/metronome/index.ts` - For context the words there will show up in the Segment field mapping UI, which looks something like this


inputs
Screen Shot 2021-11-15 at 4 44 27 PM

File: packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+   */
+  transaction_id: string
+  /**
+   * The Metronome customer ID or ingest alias this event should be associated with.
+   */
+  customer_id: string
+  /**
+   * The timestamp at which this event occurred.
+   */
+  timestamp: string | number
```
Comment (mjegilbert) [inline packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts:15]: this type seems incorrect. We only support an RFC 3999 string right?

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+              else: { '@path': '$.anonymousId' }
+            }
+          }
+        }
+      }
+    },
+    timestamp: {
+      type: 'datetime',
+      label: 'timestamp',
+      description: 'The timestamp at which this event occurred.',
```
Comment (mjegilbert) [inline packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: Should the comment include that the format is RFC 3999?

File: packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts (hunk 3 above)
Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts:15]: This is autogenerated by the fields for this action. Specifically it represents the type of the data segment gives us. Our action then needs to turn this into a format our API can accept. That happens here https://github.com/Metronome-Industries/segment-action-destinations/blob/add-metronome-destination/packages/destination-actions/src/destinations/metronome/sendEvent/index.ts#L9

and is tested here

https://github.com/Metronome-Industries/segment-action-destinations/blob/add-metronome-destination/packages/destination-actions/src/destinations/metronome/sendEvent/__tests__/index.test.ts#L55

Comment (mjegilbert) [reply packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts:15]: I see serializeEvent converts this. Do you think there might any issues with people passing in non-RFC3999 dates and getting unexpected timestamps?

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts (hunk 4 above)
Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: It should not - Since we mark the field as `date-time` in the action, segment will enforce that people only send "date-time" data to the field. The format of the data before being sent to our action is Segment's problem. 

There's a deeper discussion of this here. https://metronometechnologies.slack.com/archives/C02LRAEV4KU/p1637020717017000

I would expect at some point in the future Segment will update their actions code (in this repo) to pass in `Date` instances instead of `string | number`, and the cool thing about that is that the types will automatically fail, and they can safely fix the types and trust the test cases in that case (hopefully) without involving us at all.

File: packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts (hunk 3 above)
Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/generated-types.ts:15]: See my other comment.

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts (hunk 4 above)
Comment (mjegilbert) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: nevermind. i guess the serializeEvent will take care of this. But maybe something that says it will get converted to RFC3999?

Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: What it gets converted to doesn't really matter to the user right? As far as they know we could convert their whole event into a protobuf and send it over the wire.

Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: This field is in the Segment UI as "type: date" and then they have to select a field in their data that is a date type.

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    customer_id: {
+      type: 'string',
+      label: 'customer_id',
+      description: 'The Metronome customer ID or ingest alias this event should be associated with.',
+      required: true,
+      default: {
+        // By default, use the group ID if it exists, otherwise use the customer ID. If neither exist, use the anonymous ID.
+        '@if': {
+          exists: { '@path': '$.groupId' },
```
Comment (mjegilbert) [inline packages/destination-actions/src/destinations/metronome/sendEvent/index.ts]: Can you explain what this @path syntax is doing?
Also, how come we are looking for groupId? Where is that coming from?
The comment also suggests we are going to use the customer ID but this is looking for userId. Are those the same thing?

File: packages/destination-actions/src/destinations/metronome/sendEvent/__tests__/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+          customer_id: event.groupId || event.userId || event.anonymousId,
+          properties: event.context,
+          transaction_id: event.messageId,
+          timestamp: new Date(event.timestamp ?? "").toISOString(),
+        }
+      ])
+    })
+
+
+    it('should convert a non rfc3999 timestamp', async () => {
```
Comment (mjegilbert) [inline packages/destination-actions/src/destinations/metronome/sendEvent/__tests__/index.test.ts:55]: can you also include a test checking we convert epoch timestamps properly too?

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts (hunk 4 above)
Comment (mjegilbert) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: yeah that makes sense. My original concern was that it would be hard for clients to figure out what the converted value would be (if they didnt pass in a full rfc3999 string) but I think this is actually a moot point. Doing a few conversions using Date() everything converts as expected with the toISOString() method.

File: packages/destination-actions/src/destinations/metronome/sendEvent/__tests__/index.test.ts (hunk 6 above)
Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/__tests__/index.test.ts:55]: 4c435e8

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts (hunk 5 above)
Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts]: Idk what the @ path syntax is besides being how other action destinations work. Basically it allows us set smart defaults for the mapping the user will do between the segment event https://segment.com/docs/connections/spec/common/ and our fields. It's not documented but there _are_ TS types for it and all the other destinations use it, so I assume documentation will come.

`groupId` from Segment is described as 

> Group / Account ID. 
> This is useful in B2B use cases where you need to attribute your non-group calls to a company or account. It is relied on by several Customer Success and CRM tools.

and 

> The group API call is how you associate an individual user with a group—be it a company, organization, account, project, team or whatever other crazy name you came up with for the same concept!

Basically, for the ideal Segment user, the `userId` would be the ID of the person taking the action, and the `groupId` would the the team or org they're in. `anonymousId` is some hopefully persistent ID that could change at any time. 

As such, we use those fields (in order) as the Metronome customer ID, since the order roughly reflects how useful they'd be to us. 

This is just a default for the UI, when setting up Metronome in Segment you'd have to explicitly choose how you want to populate `customer_id` given your data, so it doesn't matter if this isn't exactly right - It's expected that users choose the right setting here, this is just meant to be a helpful default.

File: packages/destination-actions/src/destinations/metronome/sendEvent/index.ts (hunk 4 above)
Comment (cozmo) [reply packages/destination-actions/src/destinations/metronome/sendEvent/index.ts:59]: I switched to `dayjs` which everyone thinks should be the same as `new Date()` but segment has only promised to pass things that match what `dayjs` expects so might as well be safe. Also added a comment to this point.

**Review by mjegilbert — APPROVED** [review]
